winsup/utils/getprocaddr.c

```diff
@@ -0,0 +1,174 @@
+
+static int pid;
+static uintptr_t exit_code;
+
+/* Avoid fprintf(), as it would try to reference '__getreent' */
+static void
+output (BOOL error, const char *fmt, ...)
+{
+  va_list ap;
+  char buffer[1024];
+
+  va_start (ap, fmt);
+  vsnprintf (buffer, sizeof(buffer) - 1, fmt, ap);
+  buffer[sizeof(buffer) - 1] = '\0';
+  va_end (ap);
+  WriteFile (GetStdHandle(error ? STD_ERROR_HANDLE : STD_OUTPUT_HANDLE),
+	      buffer, strlen (buffer), NULL, NULL);
+}
+
+static int
+inject_remote_thread_into_process(HANDLE process, LPTHREAD_START_ROUTINE address, uintptr_t exit_code)
+{
+  int res = -1;
+
+  if (!address)
+    return res;
+  DWORD thread_id;
+  HANDLE thread = CreateRemoteThread(process, NULL, 1024 * 1024, address,
+                                     (PVOID)exit_code, 0, &thread_id);
+  if (thread)
+  {
+    /*
+      * Wait up to 10 seconds (arbitrary constant) for the thread to finish;
+      * After that grace period, fall back to terminating non-gently.
+      */
+    if (WaitForSingleObject(thread, 10000) == WAIT_OBJECT_0)
+      res = 0;
+    CloseHandle(thread);
+  }
+
+  return res;
+}
+
+static WINAPI BOOL
+ctrl_handler(DWORD ctrl_type)
+{
+  unsigned short count;
+  void *address;
+  HANDLE process;
+  HANDLE main_process = OpenProcess(PROCESS_CREATE_THREAD |
+                                        PROCESS_QUERY_INFORMATION |
+                                        PROCESS_VM_OPERATION | PROCESS_VM_WRITE |
+                                        PROCESS_VM_READ,
+                                    FALSE, pid);
+  if (main_process == NULL)
+    return 1;
+  PSYMBOL_INFOW info;
+  DWORD64 displacement;
+
+  count = CaptureStackBackTrace (1l /* skip this function */,
+			         1l /* return only one trace item */,
+				 &address, NULL);
+  if (count != 1)
+    {
+      output (1, "Could not capture backtrace\n");
+      return FALSE;
+    }
+
+  process = GetCurrentProcess ();
+  if (!SymInitialize (process, NULL, TRUE))
+    {
+      output (1, "Could not initialize symbols\n");
+      return FALSE;
+    }
+
+  info = (PSYMBOL_INFOW)
+    malloc (sizeof(*info) + MAX_SYM_NAME * sizeof(wchar_t));
+  if (!info)
+    {
+      output (1, "Could not allocate symbol info structure\n");
+      return FALSE;
+    }
+  info->SizeOfStruct = sizeof(*info);
+  info->MaxNameLen = MAX_SYM_NAME;
+
+  if (!SymFromAddrW (process, (DWORD64)(intptr_t)address, &displacement, info))
+    {
+      output (1, "Could not get symbol info\n");
+      SymCleanup(process);
+      return FALSE;
+    }
+
+  /* Inject the remote thread first*/
+  inject_remote_thread_into_process (main_process, (LPTHREAD_START_ROUTINE)(intptr_t)info->Address, exit_code);
+  
+  output (0, "%p\n", (void *)(intptr_t)info->Address);
+
+  CloseHandle(GetStdHandle(STD_OUTPUT_HANDLE));
+  SymCleanup(process);
+
+  exit(0);
+}
+
+int
+main (int argc, char **argv)
+{
+  char *end;
+
+  if (argc < 4)
+    {
+      output (1, "Need a function name, exit code and pid\n");
```

Comment:
Maybe it would make more sense to allow `argc == 2`, so that we do not have to change the name of the helper (`getprocaddr` implies that we _do_ print out an address, but we inject a remote thread calling that address instead. But if we allow `argc == 2` to print out the address, and only if `argc == 4` we inject the remote thread instead, the name of the executable still makes sense).

And now that I read all this, I remember why I needed that `--alloc-console` flag: if the calling process has the same bitness as the called process, they share a console, and if a Ctrl+C event is sent to that console, all processes attached to the same console are terminated. Including the MSYS2 runtime.

So I think that we will need to be a bit more careful here: I think the safest thing would be to _always_ attach a new console if `argv[1]` is `CtrlRoutine`, i.e. right before the `if (!SetConsoleCtrlHandler (ctrl_handler, TRUE))` line.

Reply:
> And now that I read all this, I remember why I needed that --alloc-console flag: if the calling process has the same bitness as the called process, they share a console, and if a Ctrl+C event is sent to that console, all processes attached to the same console are terminated. Including the MSYS2 runtime.

Yeah, I have seen that happen multiple times and was wondering what was the problem. After, that I thought that this shouldn't be called from minitty and then used cmd.

Reply:
But this doesn't happen when I press `Ctrl+C` in the console though.

Reply:
Can you check if https://github.com/msys2/msys2-runtime/pull/31/commits/5cbbec772318bd7ef2376a6e0623eb0eeeb63c8d works?

Reply:
> 
> 
> Can you check if [5cbbec7](https://github.com/msys2/msys2-runtime/commit/5cbbec772318bd7ef2376a6e0623eb0eeeb63c8d) works?

I won't have time to actually try this, but from what I read, it looks good (although it has tons of whitespace changes, which makes the patch a bit hard to read).

Reply:
I have tested it out locally before pushing and can say it works.

> (although it has tons of whitespace changes, which makes the patch a bit hard to read).

I have no idea why this happens each time when I edit something. Maybe vscode does some formatting changes automatically?

Reply:
> I think the safest thing would be to always attach a new console if argv[1] is CtrlRoutine, i.e. right before the if (!SetConsoleCtrlHandler (ctrl_handler, TRUE)) line.

I think it is not needed, it works without that also.

Reply:
The problem is that the `GenerateConsoleCtrlEvent()` call might take down the calling process. Which is kind of counterproductive if all you want is to figure out the address of the `CtrlRoutine` function ;-)

Reply:
> Maybe vscode does some formatting changes automatically?

Probably.

Reply:
I have previously tried `--alloc-console` and it produced a popup. Do we want to produce a popup every time the user wants to kill the program(by pressing Ctrl+C)?
Or that doesn't happen when called from runtime itself?

Reply:
> Which is kind of counterproductive if all you want is to figure out the address of the CtrlRoutine function ;-)

We aren't injecting the remote thread when 2 args are passed or when want to find the address so I think we don't need to do so.

Reply:
> I have previously tried `--alloc-console` and it produced a popup

Then it uses incorrect parameters. There is a way to create a console and hide it before it is shown. I think you need to add a creation flag `CREATE_NEW_CONSOLE` and then set `STARTF_USESHOWWINDOW` in the startup info and then set `wShowWindow` to `SW_HIDE`.

> > Which is kind of counterproductive if all you want is to figure out the address of the CtrlRoutine function ;-)
> 
> We aren't injecting the remote thread when 2 args are passed or when want to find the address so I think we don't need to do so.

We are sending the control event in order to get the backtrace so that we can determine the address. We will get that address all right, but unfortunately the buck does not stop there: the control event is interpreted, and the programs attached to the same console are taken down (there seems to be some special handling for `cmd.exe` because, as you pointed out, it is not taken down by <kbd>Ctrl</kbd>+<kbd>C</kbd>).

Reply:
> Then it uses incorrect parameters. There is a way to create a console and hide it before it is shown. I think you need to add a creation flag CREATE_NEW_CONSOLE and then set STARTF_USESHOWWINDOW in the startup info and then set wShowWindow to SW_HIDE.

:+1:  Maybe I try to do it later today.

> We are sending the control event in order to get the backtrace so that we can determine the address. We will get that address all right, but unfortunately the buck does not stop there: the control event is interpreted, and the programs attached to the same console are taken down (there seems to be some special handling for cmd.exe because, as you pointed out, it is not taken down by Ctrl+C).

I have seen that happen only when we inject a remote thread not otherwise(yes there is special handling for cmd.exe even when injecting a remote thread). I was able to call `getprocaddr64` present in Git Bash without the Git Bash being closed itself.
```sh
$ /usr/libexec/getprocaddr64 CtrlRoutine
00000000772e4810
Segmentation fault
```
Anyhow I will try implementing what you told.

Reply:
I tried implementing it and I am unhappy with it. It nearly takes around 5-7 seconds extra than the previous implementation also wasn't able to use those executable on cmd(there was no output in cmd, IDK). Though I can confirm that the current implementation works without closing the original shell itself, I have tried it, on both cmd and bash.

Anyway, I attached the diff and don't want to push until you tell the `5-7` seconds extra time is ok.
[console.txt](https://github.com/msys2/msys2-runtime/files/6378723/console.txt)

Reply:
Hmm. 5-7 seconds is not good at all. But I wonder where they come from... any idea?

Reply:
> any idea?

No idea. But I can give data on how much time it really takes. here, `new.exe` is with this patch applied and `old.exe` is without this patch(same as in this PR now)

```sh
$ time ./new CtrlRoutine
0000000076e14810
Segmentation fault

real    0m10.100s
user    0m0.000s
sys     0m0.046s

$ time ./old CtrlRoutine
0000000076e14810
Segmentation fault

real    0m0.060s
user    0m0.015s
sys     0m0.015s
````

```sh
$ time ./new ExitProcess
0000000076f1c120
Segmentation fault

real    0m0.060s
user    0m0.000s
sys     0m0.015s

$ time ./old ExitProcess
0000000076f1c120
Segmentation fault

real    0m0.040s
user    0m0.000s
sys     0m0.015s
```

```sh
$ time ./old CtrlRoutine 0 9848

real    0m0.070s
user    0m0.000s
sys     0m0.031s

$ time ./new CtrlRoutine 0 10052

real    0m10.110s
user    0m0.000s
sys     0m0.031s
```

Reply:
ah, trying this in MSYS2 shell for `old.exe` simply closed the shell, but the weird thing is it works in git bash(the above time function were from Git Bash).

Reply:
@dscho Can you help me out so that I can get this PR in?

Reply:
I will try to find some time, but this week I'm a bit short on it.

Reply:
Maybe I think we can rename the executable and not worry about this much?

Reply:
I don't think we can move forward on this before we understand the root cause for those 7 seconds.

Reply:
For the record, I just compiled this and I do not see those seven seconds.

However, what I _do_ notice is that my _entire_ `tmux` session is _GONE_ when I run `getprocaddr64.exe CtrlRoutine`.

That suggests to me that my advice to put back that `--alloc-console` logic for the `CtrlRoutine` part was spot on. It is absolutely required. Without it you will annoy users, just like it annoyed me when not only the current terminal but all 17 tmux panes vanished without a trace.

Reply:
When reinstating the `--alloc-console` logic (although implicitly, when `CtrlRoutine` was asked for), I really don't see those 7 (or 10) seconds at all. And I don't see that segmentation fault, either:

```
$ time ./x86_64-pc-msys/winsup/utils/getprocaddr64.exe CtrlRoutine
00007ff8cf91b340

real    0m0.251s
user    0m0.000s
sys     0m0.046s

$ time ./x86_64-pc-msys/winsup/utils/getprocaddr64.exe CtrlRoutine
00007ff8cf91b340

real    0m0.231s
user    0m0.000s
sys     0m0.015s

$ time ./x86_64-pc-msys/winsup/utils/getprocaddr64.exe TerminateProcess
00007ff8d0d20760

real    0m0.103s
user    0m0.000s
sys     0m0.062s

$ time ./x86_64-pc-msys/winsup/utils/getprocaddr64.exe TerminateProcess
00007ff8d0d20760

real    0m0.139s
user    0m0.015s
sys     0m0.062s
```

As you can see, `CtrlRoutine` is slower than `TerminateProcess`, but that's probably to be expected.

I _could_ imagine that you are seeing a 7 second penalty because of some anti-malware that gets suspicious about the stack tracing.

For the record, I reinstated the `--alloc-console` logic thusly:

```diff
diff --git a/winsup/utils/getprocaddr.c b/winsup/utils/getprocaddr.c
index 0fa07827ec..54c252d934 100644
--- a/winsup/utils/getprocaddr.c
+++ b/winsup/utils/getprocaddr.c
@@ -179,6 +179,18 @@ int main(int argc, char **argv)
         return 1;
     }
 
+    /* Avoid terminating all processes attached to the current console */
+    if (!FreeConsole () && GetLastError () != ERROR_INVALID_PARAMETER)
+      {
+        output (1, "Could not detach from current Console: %d\n", (int)GetLastError());
+	return 1;
+      }
+    if (!AllocConsole ())
+      {
+        output (1, "Could not allocate a new Console\n");
+	return 1;
+      }
+
     if (!SetConsoleCtrlHandler(ctrl_handler, TRUE))
     {
         output(1, "Could not register Ctrl handler\n");
```

Reply:
On further thought, that comment should probably be a bit more verbose. Something like this:

```c
    /*
     * Avoid terminating all processes attached to the current console;
     * This would happen if we used the same console as the caller, though,
     * because we are sending a CtrlEvent on purpose (which _is_ sent to
     * all processes connected to the same console, and the other processes
     * are most likely unprepared for that CTRL_BREAK_EVENT and would be
     * terminated as a consequence, _including the caller_).
     */
```

Reply:
First, using the logic which uses `AllocConsole` means it going to create a popup, there was no way to customize that directly, as in the docs https://docs.microsoft.com/en-us/windows/console/allocconsole. It does create a popup when I compile and run with your patch... It doesn't look good to have a popup on pressing Ctrl+C sadly.

So, on reading https://docs.microsoft.com/en-us/windows/console/creation-of-a-console, it seems we can launch a `cmd` using `CreateProcess` and attach to it using [`AttachConsole`](https://docs.microsoft.com/en-us/windows/console/attachconsole). My patch, which I previously linked does it, but it fails somewhere. I can look into fixing it though. And, that was the reason I was seeing some extra seconds. I can possibly fix them.

Next, about SegFaults, it happens only when it print's something say a help message. For example, this happens.
```sh
$ ./getprocaddr.exe
Need a function name, exit code and pid
Segmentation fault
```

Reply:
> It does create a popup when I compile and run with your patch... It doesn't look good to have a popup on pressing Ctrl+C sadly.

Don't we already call `CreateProcessW()` with the console hidden?

*clicketyclick*

Yes we do: https://github.com/msys2/msys2-runtime/pull/31/files#diff-083ad673bed6c3e0d175243e2aa51f3c5ea6576894e2cd06b8c78f654f5da4a1R92

We even make sure to open a new process group (which I _think_ means that it creates a new console).

To make sure, we could guard [the code I mentioned above](https://github.com/msys2/msys2-runtime/pull/31#discussion_r635588030) behind a check that calls [`GetConsoleProcessList()`](https://docs.microsoft.com/en-us/windows/console/getconsoleprocesslist) and only attaches a new console if that function returned something larger than 1 (because that means that not only the current process is attached to the current console, but other processes, too).

I don't think we should call `cmd.exe` just to avoid a console from popping up, ever ;-)

Reply:
> about SegFaults, it happens only when it print's something say a help message. For example, this happens.
> 
> ```shell
> $ ./getprocaddr.exe
> Need a function name, exit code and pid
> Segmentation fault
> ```

Have you tried running it in `gdb` to figure out where that segmentation fault is coming from?

Reply:
> Have you tried running it in gdb to figure out where that segmentation fault is coming from?

I have, but strangely it doesn't segfault there :thinking: 
```sh
(gdb) run
Starting program: H:\mingw\test_programs\newgdb\getprocaddr.exe
Need a function name, exit code and pid
Or needs a function name.
[Inferior 1 (process 7460) exited with code 01]
```
Maybe it is something with output handles, not sure.

> I don't think we should call cmd.exe just to avoid a console from popping up, ever ;-)

OK, I was doing that, and that's why those 6-7 seconds :facepalm: 

> Don't we already call CreateProcessW() with the console hidden?

So, that means we will not get a popup when it is called from runtime itself?
I think for creating a new console `CREATE_NEW_CONSOLE` flag should be passed. https://docs.microsoft.com/en-us/windows/win32/procthread/process-creation-flags#flags

> To make sure, we could guard the code I mentioned above behind a check that calls GetConsoleProcessList() and only attaches a new console if that function returned something larger than 1 (because that means that not only the current process is attached to the current console, but other processes, too).

That looks good. If it called from runtime we don't need to create a console, if it is from a normal terminal we should, right?

Reply:
Edit about gdb: I used `winpty` to run `gdb` which is why I didn't get SegFault there. (it gives nice colours there, you know?)

When running plainly `gdb` in MSYS2 console, I get
```sh
(gdb) run
Starting program: H:\mingw\test_programs\newgdb\getprocaddr.exe
Need a function name, exit code and pid

Program received signal SIGSEGV, Segmentation fault.
0x000007fefd061b44 in WriteFile () from C:\Windows\system32\KernelBase.dll
(gdb) backtrace
#0  0x000007fefd061b44 in WriteFile () from C:\Windows\system32\KernelBase.dll
#1  0x00000000773c1b06 in WriteFile () from C:\Windows\system32\kernel32.dll
#2  0x000000013fc41635 in output (error=1,
    fmt=0x13fc4a0b8 "Need a function name, exit code and pid\n")
    at getprocaddr.c:37
#3  0x000000013fc41a86 in main (argc=1, argv=0x6b82b0) at getprocaddr.c:179
```

Reply:
Ok, this is done in https://github.com/msys2/msys2-runtime/pull/31/commits/ee97ce904f8a09960eb5a6904b2e1e2312d82e04

and it seems to work well, if called from console creates a popup, if called from runtime simply executes. Thanks for your help.

Reply:
> > Don't we already call CreateProcessW() with the console hidden?
> 
> So, that means we will not get a popup when it is called from runtime itself?

Exactly.

> I think for creating a new console `CREATE_NEW_CONSOLE` flag should be passed. https://docs.microsoft.com/en-us/windows/win32/procthread/process-creation-flags#flags

I think we do need CREATE_NEW_PROCESS_GROUP, quoting part from [here](https://docs.microsoft.com/en-us/windows/win32/procthread/process-creation-flags#flags):

> Process groups are used by the GenerateConsoleCtrlEvent function to enable sending a CTRL+BREAK signal to a group of console processes.

That sounds like exactly what we want.

> I used `winpty` to run `gdb` which is why I didn't get SegFault there. (it gives nice colours there, you know?)

Yes, I did know about `winpty`, it also fixes the <kbd>Tab</kbd> handling when you use the MINGW version of `gdb`. I use it all the time.

> When running plainly `gdb` in MSYS2 console, I get
> 
> ```shell
> (gdb) run
> Starting program: H:\mingw\test_programs\newgdb\getprocaddr.exe
> Need a function name, exit code and pid
> 
> Program received signal SIGSEGV, Segmentation fault.
> 0x000007fefd061b44 in WriteFile () from C:\Windows\system32\KernelBase.dll
> (gdb) backtrace
> #0  0x000007fefd061b44 in WriteFile () from C:\Windows\system32\KernelBase.dll
> #1  0x00000000773c1b06 in WriteFile () from C:\Windows\system32\kernel32.dll
> #2  0x000000013fc41635 in output (error=1,
>     fmt=0x13fc4a0b8 "Need a function name, exit code and pid\n")
>     at getprocaddr.c:37
> #3  0x000000013fc41a86 in main (argc=1, argv=0x6b82b0) at getprocaddr.c:179
> ```

What a bummer. That still does tell me what is going wrong, but it does suggest that we should switch from using `output()` directly to `printf()` for `output(0, ...)` calls, and to `fprintf(stderr, ...)` for `output(1, ...)` calls.

> Ok, this is done in [ee97ce9](https://github.com/msys2/msys2-runtime/commit/ee97ce904f8a09960eb5a6904b2e1e2312d82e04)

👍 

> and it seems to work well, if called from console creates a popup, if called from runtime simply executes. Thanks for your help.

If you want to avoid that popup, you could probably use `IsWindowVisible(GetConsoleWindow())` to check whether the current console window is shown, and re-execute (using `_wpgmptr` as `lpApplicationName` and `GetCommandLineW()` as `lpCommandLine` in a `CreateProcessW()` call, and of course setting the flags so that the console window is hidden).

I am not really sure this is worth the effort, though, because this is an executable _meant_ to be called from the MSYS2 runtime, after all.

Reply:
> I think we do need CREATE_NEW_PROCESS_GROUP, quoting part from here:

:+1: 

> What a bummer. That still does tell me what is going wrong, but it does suggest that we should switch from using output() directly to printf() for output(0, ...) calls, and to fprintf(stderr, ...) for output(1, ...) calls.

I still think, "the pointer"(as explained [here](https://github.com/msys2/msys2-runtime/pull/31#discussion_r636942164)) doesn't print when using `fprintf`, so we should use this output function there.

> I am not really sure this is worth the effort, though, because this is an executable meant to be called from the MSYS2 runtime, after all.

yes, so we don't need to worry about that.

Reply:
> > What a bummer. That still does tell me what is going wrong, but it does suggest that we should switch from using output() directly to printf() for output(0, ...) calls, and to fprintf(stderr, ...) for output(1, ...) calls.
> 
> I still think, "the pointer"(as explained [here](https://github.com/msys2/msys2-runtime/pull/31#discussion_r636942164)) doesn't print when using `fprintf`, so we should use this output function there.

Oh, you mean that `stderr` and `stdout` now use the allocated console... that makes sense. But then we probably want to document that idea above the `output()` function.

Or (and this would be my preferred solution) we could ensure that the `stderr` and `stdout` are not redirected to the allocated console, via something like this:

```c
  if (GetConsoleProcessList (&proc_lists, 5) > 1)
    {
      HANDLE out = GetStdHandle (STD_OUTPUT_HANDLE);
      HANDLE err = GetStdHandle (STD_ERROR_HANDLE);

      if (!FreeConsole () && GetLastError () != ERROR_INVALID_PARAMETER)
        {
          fprintf (stderr, "Could not detach from current Console: %d\n",
                   (int)GetLastError ());
          return 1;
        }
      if (!AllocConsole ())
        {
          fprintf (stderr, "Could not allocate a new Console\n");
          return 1;
        }
      SetStdHandle (STD_OUTPUT_HANDLE, out);
      SetStdHandle (STD_ERROR_HANDLE, err);
    }
```

(assuming that that works).

Reply:
I will look into it later today.

Reply:
I think that doesn't my patch looks like
```patch
diff --git a/winsup/utils/getprocaddr.c b/winsup/utils/getprocaddr.c
index ed0083c39..cde8bee2b 100644
--- a/winsup/utils/getprocaddr.c
+++ b/winsup/utils/getprocaddr.c
@@ -202,6 +202,9 @@ main (int argc, char **argv)
   DWORD proc_lists;
   if (GetConsoleProcessList (&proc_lists, 5) > 1)
     {
+      HANDLE out = GetStdHandle (STD_OUTPUT_HANDLE);
+      HANDLE err = GetStdHandle (STD_ERROR_HANDLE);
+
       if (!FreeConsole () && GetLastError () != ERROR_INVALID_PARAMETER)
         {
           fprintf (stderr, "Could not detach from current Console: %d\n",
@@ -213,6 +216,8 @@ main (int argc, char **argv)
           fprintf (stderr, "Could not allocate a new Console\n");
           return 1;
         }
+      SetStdHandle (STD_OUTPUT_HANDLE, out);
+      SetStdHandle (STD_ERROR_HANDLE, err);
     }

   if (!SetConsoleCtrlHandler (ctrl_handler, TRUE))
@@ -242,4 +247,5 @@ main (int argc, char **argv)
       fprintf (stderr, "WaitForSingleObject failed (%d)\n", GetLastError ());
       return 1;
     }
+  return 0;
 }
```
and I don't print the pointer both on bash and cmd.

Reply:
Hmm, since the one thing that prints after allocating a new console is just the pointer how about using the same `output` function there?

Reply:
I really like the ease of `fprintf()`, so I'd like to make it work _somehow_. I'll look into it later today, if I find the time.

Reply:
Ok. I would be busy for the next few days, only after that, I can look into this.

Reply:
So... it looks as if I do not actually need that `GetStdHandle()`/`SetStdHandle()` dance, but instead I only need to `fflush (stdout)` (I ran into this issue a couple of times in the past, where the standard handles of non-MSYS2 processes were connected to MSYS2's named pipes which needed _explicit_ flushing, even for `stderr`).

Therefore, this diff works for me:

```diff
diff --git a/winsup/utils/getprocaddr.c b/winsup/utils/getprocaddr.c
index bddad7004f..6783bc59ad 100644
--- a/winsup/utils/getprocaddr.c
+++ b/winsup/utils/getprocaddr.c
@@ -97,6 +97,7 @@ ctrl_handler (DWORD ctrl_type)
   if (pid == 0)
     {
       printf ("%p\n", (void *)(intptr_t)info->Address);
+      fflush (stdout);
     }
   else
     {
@@ -166,6 +167,7 @@ main (int argc, char **argv)
       if (pid == 0)
         {
           printf ("%p\n", address);
+          fflush (stdout);
           return 0;
         }
       HANDLE main_process = OpenProcess (
```

Reply:
Ah, nice, this seems to work at least in MSYS2 shell. So, I would apply this patch. But this doesn't seem to work when running with CMD though (I think it's fine to ignore).

Reply:
Oh, I _think_ this might be our own fault:

```
  CloseHandle (GetStdHandle (STD_OUTPUT_HANDLE));
```

is called in `ctrl_handler()`, without flushing first.

Reply:
Yep, in my testing, removing this statement fixed the issue!

FWIW this statement was necessary back when we read the output from the spawned process in the MSYS2 runtime _and then_ injected the remote thread _in the runtime_. It was necessary because we needed to flush things and finalize that particular connection so that the spawned process would actually exit.

Reply:
> Yep, in my testing, removing this statement fixed the issue!

Hmm, it still doesn't work for me in cmd (got the latest build from CI), but I think that shouldn't be a problem though.

Reply:
Hrm. I guess we still have to call `fflush(stdout);` manually?